using_openshift/port_forwarding.adoc

```diff
@@ -0,0 +1,86 @@
+= Port Forwarding
+{product-author}
+{product-version}
+:data-uri:
+:icons:
+:experimental:
+:toc: macro
+:toc-title:
+
+toc::[]
+
+== Overview
+OpenShift takes advantage of a feature built into Kubernetes to support port
+forwarding to pods. This is implemented using HTTP along with a multiplexed
+streaming protocol such as link:http://www.chromium.org/spdy[SPDY] or
+link:https://http2.github.io/[HTTP/2].
+
+== Client Operation
+Support for port forwarding is
+link:../using_openshift/cli.html#common-cli-operations[built into the CLI]:
+
+****
+`$ osc port-forward -p _<pod>_ _[<local port:]<pod port>_ _[[<local port:]<pod port> ...]_`
+****
+
+////
+The CLI listens on each local port specified by the user, forwarding via the
+link:#protocol[protocol] described below.
+////
+
+Ports may be specified using the following formats:
+
+`5000`::
+  The client listens on port 5000 locally and forwards to 5000 in the
+pod.
+`6000:5000`::
+  The client listens on port 6000 locally and forwards to 5000
+in the pod.
+`:5000` or `0:5000`::
+  The client selects a free local port and
+forwards to 5000 in the pod.
+
+////
+[protocol]
+*Protocol*
+////
+
+Clients initiate port forwarding to a pod by issuing a request to the
+Kubernetes API server:
+
+----
+/proxy/minions/node123.openshift.com/portForward/myns/mypod
+----
+
+In the above URL:
+
+* `_node123.openshift.com_` is the node name.
+* `_myns_` is the namespace of the target pod.
+* `_mypod_` is the name of the target pod.
+
+After sending a port forward request to the API server, the client upgrades the
+connection to one that supports multiplexed streams; the current implementation
+uses *SPDY*.
+
+The client creates a stream with the `port` header containing the target port in
+the pod. All data written to the stream is delivered via the Kubelet to the
+target pod and port. Similarly, all data sent from the pod for that forwarded
+connection is delivered back to the same stream in the client.
+
+The client closes all streams, the upgraded connection, and the underlying
+connection when it is finished with the port forwarding request.
+
+== Server Operation
+The Kubelet handles port forward requests from clients. Upon receiving a
+request, it upgrades the response and waits for the client to create port
+forwarding streams. When it receives a new stream, it copies data between the
+stream and the pod's port.
+
+*Forwarding to a Pod's Port*
+
+There are a few different options for forwarding to a pod's port. The current
+implementation invokes `nsenter` directly on the node host to enter the pod's
+network namespace and invoke `socat` to copy data between the stream and the
+pod's port. Other options include running a "helper" pod that then runs
+`nsenter` and `socat`, so those binaries are not required to be installed on the
+host.
```

Comment:
The reason I had mentioned multiple options for port forwarding was because the doc was originally somewhat of a combination of "how an end user does port forwarding" and "technical implementation details", with the options targeting people interested in how it's built and what architectural decisions we had made.

We only have 1 implementation now, using nsenter and socat. There are no choices or options at this point.